04-onion-routing.md

```diff
@@ -1407,6 +1438,88 @@ The _origin node_:
+  resolution (`update_fulfill_hltc` / `update_fail_hltc` / 
+  `update_fail_malformed_htlc`).
+
+Each HTLC's contribution to reputation is expressed by its `effective_fees` 
+which capture the fees that it paid and the opportunity cost that its 
+`resolution_time` incurred. 
+- `fees`: the fees paid by a forwarded HTLC (as described in [BOLT #7](07-routing-gossip.md#htlc-fees), 
+  equal to 0 if the HTLC was failed).
+- `opportunity_cost`: 'ceil ( (resolution_time - resolution_period) / resolution_period) * fees`
+
+Given that `resolution_time` will be > 0 in practice, `opportunity_cost` is 0 
+for HTLCs that resolve within `resolution_period`, and charges the `fees` that 
+the HTLC would have earned per period it is held thereafter. This cost accounts 
+for the slot and liquidity that could have otherwise been paid for by 
+successful, fast resolving HTLCs during the `resolution_time` the HTLC was 
+locked in the channel.
+
+For every incoming HLTC a peer has forwarded through a node, its `effective_fees` 
+are calculated as follows:
+- if `endorsed` = 1 in `update_add_htlc`:
+  - `effective_fees` = `fees` - `opportunity_cost` 
+- otherwise: 
+  - if successfully resolved AND `resolution_time` <= `resolution_period`
+    - `effective_fees` = `fees`
+  - otherwise: 
+    - `effective_fees` = 0
+
+The `effective_fees` that a peer has paid our node are compared to our total 
+routing revenue to classify a peer's reputation as "good" or "neutral". This
+relates the fees that must be paid to earn "good" reputation to the damage that 
+can be done by abusing it.
+
+Routing revenue is assessed over a sliding window, so that reputation 
+classification is related to the node's current routing activity. Nodes MAY
+choose a window per their preferences - shorter windows being more reactive to 
+recent routing patterns, and longer windows aggregating trends over time.
+
+- `routing_window`: the period of time in which total routing revenue is 
+  assessed (default: 10 minutes * the maximum number of blocks in the future a
+  HTLC's expiry height can fall before the node will send `expiry_too_far`, as 
+  outlined in [Failure Messages](#failure-messages)).
+- `routing_revenue`: the sum of all fees paid to the node to forward HTLCs
+  over the interval [ `now` - `routing_window` ; `now` ].
+
+The total `effective_fees` that an individual peer has generated are assessed
+over a longer period of time to relate its reputation classification to the 
+routing activity of the node (represented by `routing_revenue`).
+
+- `reputation_window`: the period of time over which the node's `effective_fees`
+  are calculated to classify its reputation (recommended default = 10 * `routing_window`).
+- `reputation_revenue`: the total `effective_fees` of HTLCs forwarded by the peer
+  over the interval [ `now` - `reputation_window` ; `now` ].
+
+A peer is classified as having "good" local reputation iff `reputation_revenue` >= `routing_revenue`, 
```

Comment:
So IIUC everytime an `unkown` HTLC is fulfilled it **may** add to the `effective_fees` but will **always** add to `routing_revenue` for a given window. Given that the network upgrades to this reputation system and all `reputation_revenue` is initially `0` & with HTLCs effectively being `unkown`, how would peers get out of this entropy where `routing_revenue` will always be greater than `reputation_revenue`?

An attempt to make a more simple rephrase of the above question:
*When will the ever-first `endorsed` HTLC occur in the network?*

Reply:
> `routing_revenue` is calculated over a shorter period of time than `repuation_revenue`, so nodes have an opportunity to build reputation with consistent behaviour.

Oh, important detail that I missed! This also eliminates some of my comments [here](https://github.com/lightning/bolts/pull/1071#discussion_r1195682786), as there seems to be a native tention to push reputation up slowly over time :+1: 

> If anything, we start in the opposite situation. Without any endorsement signal on any historical HTLCs, nodes will start out with their history only contributing to reputation (ie, fast-resolving, successful HTLCs). I think this is correct, because the current state of LN is essentially "everyone trusts everyone", and then we gradually update to be more discerning.

Ok, so after upgrading `routing_revenue` will be >=0 but `reputation_revenue` will be initialized to 0. Nodes would start to send out non-endorsed HTLCs and start building up `reputation_revenue`, becoming "good" peers as soon as `routing_window` starts rotating (as `reputation_revenue` will remain for 10 times the window duration).

Reply:
Right, routing history will still be there to make the calculations on the routing/revenue windows :+1: 

ty for explaining